tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -1387,7 +1417,27 @@ public:
 
    void Initialize()
    {
+      ::TDirectory::TContext c; // do not change gDirectory outside of this scope
+
       const auto cs = ROOT::CompressionSettings(fOptions.fCompressionAlgorithm, fOptions.fCompressionLevel);
+
+      // Write an empty TTree to file as a clean starting point for TBufferMerger.
+      // This is mostly useful when a TTree with the same name is already present in the output
+      // file and we are UPDATE'ing the TFile, in which case TBufferMerger would try to merge new
```

Comment:
Should we execute this code only in the update mode case?

Reply:
It doesn't hurt to always execute it. This way the logic is always the same (easier to read, understand, follow) and I don't have to do an awkward "check if the file is there, if yes check if the tree is there".

Reply:
I was proposing "if fOptions.Mode is update then", this save on the spurious opening of file . 

Actually isn't that function also executed if the file is read-only? (and Snapshot is not called)

Reply:
There is another very good reason to always write the header: so if no event passes the selections, the file still contains an empty TTree with the proper structure. Before this change, the file would contain a TTree key with no branch information.

> Actually isn't that function also executed if the file is read-only? (and Snapshot is not called)

No, `SnapshotHelperMT::Initialize` is called right before an event loop in which a `Snapshot` has been booked.

Reply:
> No, SnapshotHelperMT::Initialize i

That this Initialize function is part of a class name SnapshotHelperMT was hard to see from github as they are many classes in the same file ... and the github expand-this-code-snippet button can expand below the snippet not above it :( .....

> There is another very good reason to always write the header: 

humm .. if the fOptions.Mode is NEW then the 2nd use later on will be wrong (lead to a Zombie file).  if the option is RECREATE then the work done here will be throw out.   So if you can going this way, then the 2nd opening of the output file must be a "UPDATE".

Reply:
> if you can going this way, then the 2nd opening of the output file must be a "UPDATE".

yes the current logic is definitely not ok :sweat: work in progress